elements/pl-units-input/pl-units-input.py

```diff
@@ -0,0 +1,468 @@
+from enum import Enum
+from html import escape
+
+import chevron
+import lxml.html
+import prairielearn as pl
+import unit_utils as uu
+from pint import Unit, errors
+from typing_extensions import assert_never
+
+
+class DisplayType(Enum):
+    INLINE = "inline"
+    BLOCK = "block"
+
+
+class GradingMode(Enum):
+    ONLY_UNITS = 1
+    EXACT_UNITS = 2
+    WITH_UNITS = 3
+
+
+GRADING_MODE_DEFAULT = GradingMode.WITH_UNITS
+WEIGHT_DEFAULT = 1
+CORRECT_ANSWER_DEFAULT = None
+LABEL_DEFAULT = None
+SUFFIX_DEFAULT = None
+DISPLAY_DEFAULT = DisplayType.INLINE
+ALLOW_BLANK_DEFAULT = False
+COMPARISON_DEFAULT = uu.ComparisonType.RELABS
+RTOL_DEFAULT = 1e-2
+ATOL_DEFAULT = "1e-8"
+DIGITS_DEFAULT = 2
+SIZE_DEFAULT = 35
+SHOW_HELP_TEXT_DEFAULT = True
+MAGNITUDE_PARTIAL_CREDIT_DEFAULT = None
+SHOW_FEEDBACK_DEFAULT = True
+
+UNITS_INPUT_MUSTACHE_TEMPLATE_NAME = "pl-units-input.mustache"
+
+
+def prepare(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    required_attribs = ["answers-name"]
+    optional_attribs = [
+        "weight",
+        "correct-answer",
+        "label",
+        "suffix",
+        "display",
+        "allow-blank",
+        "blank-value",
+        "grading-mode",
+        "comparison",
+        "rtol",
+        "atol",
+        "digits",
+        "size",
+        "show-help-text",
+        "placeholder",
+        "magnitude-partial-credit",
+        "show-feedback",
+    ]
+    pl.check_attribs(element, required_attribs, optional_attribs)
+
+    allow_blank = pl.get_string_attrib(element, "allow-blank", ALLOW_BLANK_DEFAULT)
+    if allow_blank and not pl.has_attrib(element, "blank-value"):
+        raise ValueError(
+            'Attribute "blank-value" must be provided if "allow-blank" is enabled.'
+        )
+
+    name = pl.get_string_attrib(element, "answers-name")
+    correct_answer_html = pl.get_string_attrib(
+        element, "correct-answer", CORRECT_ANSWER_DEFAULT
+    )
+
+    if correct_answer_html is not None:
+        if name in data["correct_answers"]:
+            raise ValueError(f"Duplicate correct_answers variable name: {name}")
+        data["correct_answers"][name] = correct_answer_html
+
+    correct_answer = data["correct_answers"].get(name)
+
+    digits = pl.get_integer_attrib(element, "digits", None)
+    if digits is not None and digits <= 0:
+        raise ValueError(
+            f"Number of digits specified must be at least 1, not {digits}."
+        )
+
+    atol = pl.get_string_attrib(element, "atol", ATOL_DEFAULT)
+    ureg = pl.get_unit_registry()
```

Comment:
This is just how the library is set up. It always requires a registry, and the reason it's slow and needs caching is because the default constructor reads from a file.

Reply:
Oh, I see from the docs it has something to do with speeding up operations on units. That's annoying but I guess we just deal with it.